File: packages/tale_knowledge/src/tale_knowledge/extraction/xlsx.py
"""XLSX text extraction using openpyxl.

Converts spreadsheet rows into plain text for indexing.
"""

from io import BytesIO

from loguru import logger
from openpyxl import load_workbook


async def extract_text_from_xlsx_bytes(
    xlsx_bytes: bytes,
    filename: str = "spreadsheet.xlsx",
) -> tuple[str, bool]:
    """Extract text from XLSX bytes.

    Args:
        xlsx_bytes: Raw XLSX bytes.
        filename: Filename for logging.

    Returns:
        Tuple of (extracted_text, vision_was_used). Vision is never used for XLSX.
    """
    logger.info(f"Processing XLSX: {filename}")

    wb = load_workbook(BytesIO(xlsx_bytes), read_only=True, data_only=True)
    sheets_text: list[str] = []

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        rows_text: list[str] = []

        for row in ws.iter_rows(values_only=True):
            cells = [str(cell) if cell is not None else "" for cell in row]
            line = " | ".join(cells).strip()
            if line and line != "|":
                rows_text.append(line)

        if rows_text:
            sheet_header = f"--- Sheet: {sheet_name} ---"
            sheets_text.append(f"{sheet_header}\n" + "\n".join(rows_text))

    wb.close()

    combined_text = "\n\n".join(sheets_text)
    logger.info(
        f"XLSX processing complete: {len(sheets_text)} sheets, {len(combined_text)} chars"
    )

    return combined_text, False

File: services/platform/backend/core/lib/knowledge/extraction/xlsx.ts
'use node';

/**
 * XLSX text extraction using SheetJS.
 *
 * Converts spreadsheet rows into plain text for indexing.
 */

import * as XLSX from 'xlsx';

/** Render a SheetJS cell value (string/number/boolean/Date/null) as text. */
function stringifyCell(cell: unknown): string {
  if (cell === null || cell === undefined) {
    return '';
  }
  if (typeof cell === 'string') {
    return cell;
  }
  if (
    typeof cell === 'number' ||
    typeof cell === 'boolean' ||
    typeof cell === 'bigint'
  ) {
    return cell.toString();
  }
  if (cell instanceof Date) {
    return cell.toISOString();
  }
  return '';
}

/**
 * Extract text from XLSX bytes. Returns `[text, visionUsed]`; vision is never
 * used for spreadsheets. Throws on an invalid/corrupt workbook.
 */
export async function extractTextFromXlsxBytes(
  xlsxBytes: Uint8Array,
  _filename = 'spreadsheet.xlsx',
): Promise<[string, boolean]> {
  let workbook: XLSX.WorkBook;
  try {
    workbook = XLSX.read(xlsxBytes, { type: 'array' });
  } catch (err) {
    throw new Error('Invalid or corrupt file', { cause: err });
  }

  const sheetsText: string[] = [];

  for (const sheetName of workbook.SheetNames) {
    const sheet = workbook.Sheets[sheetName];
    const rows = XLSX.utils.sheet_to_json<unknown[]>(sheet, {
      header: 1,
      blankrows: false,
      defval: null,
    });

    const rowsText: string[] = [];
    for (const row of rows) {
      const cells = row.map(stringifyCell);
      if (!cells.some((cell) => cell.trim().length > 0)) {
        continue;
      }
      const line = cells.join(' | ').trim();
      if (line) {
        rowsText.push(line);
      }
    }

    if (rowsText.length > 0) {
      sheetsText.push(`--- Sheet: ${sheetName} ---\n${rowsText.join('\n')}`);
    }
  }

  return [sheetsText.join('\n\n'), false];
}
